server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1911,6 +1930,60 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
             XContentParserUtils.ensureExpectedToken(XContentParser.Token.END_OBJECT, parser.nextToken(), parser);
             return builder.build();
         }
+
+        private IndexMetadata reuseMappings(IndexMetadata indexMetadata) {
+            if (indexMetadata.mapping() == null) {
+                return indexMetadata;
+            }
+
+            String digest = indexMetadata.mapping().getSha256();
+            MappingMetadata entry = cache.get(digest);
+            if (entry != null) {
+                IndexMetadata.Builder imBuilder = new IndexMetadata.Builder(indexMetadata);
+                imBuilder.putMapping(entry);
+                return imBuilder.build();
```

Comment:
This builder is fairly expensive if I remember correctly with all the setting lookup and the like. I wonder if we could just make this a copy constructor that swaps out the mapping metadata only like we did elsewhere and ourselves the trouble of setting up a builder.
Also, in this case you could make it transparent easily, and not create a new instance even in case the mappings are already instance equal?

Reply:
Something like this: 4bc2229d6139e7c1e2da1a9b830ea8fc7988e63d ?
I think we need to check whether the instances are not the same and I think here instances are always different?